### PR TITLE
Rewrite the configuration package (very basic for now)

### PR DESCRIPTION
**Pull request by jcxldn**

### CODE REVIEW

**Review by kokumaji — APPROVED** [review]
Maybe also add something like `YamlProvider#reload()`? Other than that, lgtm 👍 
Waiting for opinions from others

File: common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/Configuration.java [hunk 1]
```diff
@@ -4,90 +4,22 @@
...
-     * @return Options for this configuration
+     * Gets the given string from the path.
+     * @param path The path of the string to get
+     * @return The requested string, or null.
      */
-    @NotNull
-    public ConfigurationOptions options();
-}
+    public String getString(String path);
+}
```
Comment (kaylendog) [inline common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/Configuration.java:25]:
```suggestion
}

```

**Review by kaylendog — CHANGES_REQUESTED** [review]
Can we move this into `config`? Better integrates with future plans for a configuration module. Otherwise LGTM :3

**Review by ZachyFoxx — APPROVED** [review]
LGTM

File: bukkit/src/main/java/com/dumbdogdiner/stickyapi/bukkit/util/StartupUtil.java [hunk 2]
```diff
@@ -62,7 +63,15 @@ public static LocaleProvider setupLocale(@NotNull JavaPlugin plugin, @Nullable L
                     .severe("Failed to configure default locale file - perhaps you deleted it? Will create a new one.");
             // FIXME: This is horrible and needs to be improved
             try {
-                localeProvider.writeLocaleStream(plugin.getResource("messages.en_us.yml"), "messages.en_us.yml", true);
+                // Step 1:Load the yml from plugin.getResource (the internal .jar resource)
+                YamlProvider embeddedMessagesResource = new YamlProvider(plugin.getResource("messages.en_us.yml"));
```
Comment (kaylendog) [inline bukkit/src/main/java/com/dumbdogdiner/stickyapi/bukkit/util/StartupUtil.java:75]: Maybe do an explicit check to see if this exists before copying it? Altho it is kinda covered by the `e.printStackTrace()` below...

File: bungee/src/main/java/com/dumbdogdiner/stickyapi/bungeecord/util/StartupUtil.java [hunk 3]
```diff
@@ -69,8 +70,17 @@ public static LocaleProvider setupLocale(@NotNull Plugin plugin, @Nullable Local
                     .severe("Failed to configure default locale file - perhaps you deleted it? Will create a new one.");
             // FIXME: This is horrible and needs to be improved
             try {
-                localeProvider.writeLocaleStream(plugin.getResourceAsStream("messages.en_us.yml"), "messages.en_us.yml",
-                        true);
+                // Step 1:Load the yml from plugin.getResource (the internal .jar resource)
+                YamlProvider embeddedMessagesResource = new YamlProvider(plugin.getResourceAsStream("messages.en_us.yml"));
```
Comment (kaylendog) [inline bungee/src/main/java/com/dumbdogdiner/stickyapi/bungeecord/util/StartupUtil.java:82]: See above.

File: common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/FileConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2020-2021 DumbDogDiner <dumbdogdiner.com>. All rights reserved.
+ * Licensed under the MIT license, see LICENSE for more information...
+ */
+package com.dumbdogdiner.stickyapi.common.configuration;
```
Comment (kaylendog) [inline common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/FileConfiguration.java:5]: Were we not gonna move this to `stickyapi.config` module?

File: common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/providers/YamlProvider.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        this.inputStream = stream;
+        this.reload();        
+    }
+
+    public YamlProvider(File file) throws FileNotFoundException {
+        this(new FileInputStream(file));
+    }
+
+    public void reload() {
+        // Attempt to (re)load the stream
```
Comment (kaylendog) [inline common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/providers/YamlProvider.java:40]: Interesting formatting going on here - we gonna have spotless, right?
```suggestion
        // Attempt to (re)load the stream
        Yaml yaml = new Yaml();
        Map<String, Object> obj = yaml.load(this.inputStream);
        this.data = obj;
```

File: common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/providers/YamlProvider.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+    @Override
+    public boolean save(String path) {
+        FileWriter fileWriter;
+        try {
+            // Attempt to create the FileWriter object (can throw IOException)
+            fileWriter = new FileWriter(path);
+
+            // Save the config
+            save(fileWriter);
```
Comment (kaylendog) [inline common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/providers/YamlProvider.java]: For clarity's sake
```suggestion
            this.save(fileWriter);
```

File: common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/providers/YamlProvider.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        Yaml yaml = new Yaml();
+
+        Map<String, Object> obj = yaml.load(this.inputStream);
+
+        this.data = obj;
+    }
+
+	@Override
+	public String getString(String path, String def) {
+        Object value = data.get(path);
```
Comment (kaylendog) [inline common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/providers/YamlProvider.java]:
```suggestion
        Object value = this.data.get(path);
```

File: common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/providers/YamlProvider.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        return (value != null) ? value.toString() : def;
+	}
+
+	@Override
+	public String getString(String path) {
+        Object value = data.get(path);
```
Comment (kaylendog) [inline common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/providers/YamlProvider.java]:
```suggestion
        Object value = this.data.get(path);
```

File: common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/providers/YamlProvider.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    public boolean save(File output) {
+        FileWriter fileWriter;
+        try {
+            // Attempt to create the FileWriter object (can throw IOException)
+            fileWriter = new FileWriter(output);
+
+            // Save the config
+            save(fileWriter);
```
Comment (kaylendog) [inline common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/providers/YamlProvider.java]:
```suggestion
            this.save(fileWriter);
```

File: common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/providers/YamlProvider.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+import com.dumbdogdiner.stickyapi.common.configuration.FileConfiguration;
+
+import org.apache.commons.lang.Validate;
+import org.yaml.snakeyaml.DumperOptions;
+import org.yaml.snakeyaml.Yaml;
+
+public class YamlProvider implements FileConfiguration {
+
+    private Map<String, Object> data;
```
Comment (kaylendog) [inline common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/providers/YamlProvider.java]: Worth making this a concurrent hash map? Could potentially be accessed by different threads at the same time.

File: common/src/test/java/com/dumbdogdiner/stickyapi/common/configuration/providers/YamlProviderTest.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+import com.dumbdogdiner.stickyapi.common.configuration.FileConfiguration;
+import com.google.common.io.Files;
+
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+import org.junit.jupiter.api.TestInstance.Lifecycle;
+
+@TestInstance(Lifecycle.PER_CLASS)
+public class YamlProviderTest {
```
Comment (kaylendog) [inline common/src/test/java/com/dumbdogdiner/stickyapi/common/configuration/providers/YamlProviderTest.java:30]: As mentioned above, a test that forces a concurrent read/write of the config to ensure that behaves as expected.

File: common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/FileConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.io.File;
+import java.io.FileWriter;
+
+/**
+ * Interface for File-based Configurations.
+ */
+public interface FileConfiguration extends Configuration {
+
+    public void reload();
```
Comment (kaylendog) [inline common/src/main/java/com/dumbdogdiner/stickyapi/common/configuration/FileConfiguration.java:20]: Javadoc :3

**Review by kaylendog — CHANGES_REQUESTED** [review]
LGTM, minor changes :3

**Pull request merged**